File: day10/solution.hs
{-# LANGUAGE NamedFieldPuns, ScopedTypeVariables #-}

import Data.List (tails, scanl', delete, sort)
import Control.Monad (guard)
import Debug.Trace (traceShowId, traceShow)

main :: IO ()
main = do
  input <- words <$> readFile("input.txt")
  let adapters :: [Int] = read <$> input

  putStrLn "Part 1:"
  let sorted = sort $ (0: adapters)
  let diffs = 3 : zipWith (-) (tail sorted) sorted
  let ones = length $ filter (== 1) diffs
  let threes = length $ filter (== 3) diffs
  putStrLn $ show $ ones * threes

  putStrLn "Part 2:"
  putStrLn $ show $ count adapters 0

canConnect :: [Int] -> Int -> [Int]
canConnect available prev = filter (\a -> (a - prev) `elem` [1, 2, 3]) available

count :: [Int] -> Int -> Int
count available prev =
  let
    results :: [Int] = count' <$> [0..]
    goal = maximum available
    count' :: Int -> Int
    count' prev =
      if prev == goal then 1 else
        sum $ (\a -> results !! a) <$> canConnect available prev
  in
    results !! prev

File: day10/solution.py


with open('input.txt', 'r') as f:
    adapters = [int(w) for w in f.read().split()]

goal = max(adapters) + 3

def can_connect(prev):
    return [
        a for a in adapters
        if a in [prev + 1, prev + 2, prev + 3]
    ]


print(adapters)
print(can_connect(0))
print(goal)



memo = {}
def chain(prev):
    if prev in memo:
        return memo[prev]
    else:
        options = can_connect(prev)
        if not options:
            if goal - prev <= 3:
                result = 1
            else:
                result = 0
        else:
            result = sum(chain(option) for option in options)
        memo[prev] = result
        return result


print("Part 2:")
print(chain(0))
